### PR TITLE
virtio/serial: initial CONSOLE support

### PR DESCRIPTION
**Pull request by yf13**
## Summary

This adds initial support to allow virtio serial device being used as console on QEMU `rv-virt` device. It includes the following changes:

- virtio/serial: add `VIRTIO_SERIAL_CONSOLE` Kconfig and support.
- qemu-rv_start: guard u16550 uses with `HAVE_16550_CONSOLE`.
- rv-virt_appinit: probe virtio devices upon BOARD_EARLY_INIITIALIZE

## Impacts

None

## Testing

- Local checks on `rv-virt`
- CI checks

### CODE REVIEW

File: drivers/virtio/Kconfig [hunk 1]
```diff
@@ -74,6 +74,14 @@ if DRIVERS_VIRTIO_SERIAL
 config DRIVERS_VIRTIO_SERIAL_BUFSIZE
 	int "Virtio serial driver buffer size"
 	default 256
+
+config VIRTIO_SERIAL_CONSOLE
+	bool "Virtio serial console"
+	default n
+	select SERIAL_CONSOLE
+	---help---
+		This enables using first virtio serial device as console
```
Comment (acassis) [inline drivers/virtio/Kconfig:83]: Please include the information that it is output only

**Comment by acassis** [comment]
@yf13 please include Documentation/

**Comment by yf13** [comment]
> @yf13 please include Documentation/
Will do that later with a concrete config. I want use this chance to learn more about the virtio-serial driver.

**Comment by yf13** [comment]
> @yf13 please include Documentation/

Will do that later with a concrete config. I want use this chance to learn more about the virtio-serial driver.

**Comment by yf13** [comment]
> @yf13 please include Documentation/

@acassis, yes will do that after a concrete config is available, so far I want take this chance to review the basic changes and make sure I use virtio serial driver correctly.

**Comment by yf13** [comment]
@acassis, it seems that I've found the way to use virtio-serial as normal console now, so we no longer need the `output only` warnings. I still need time to clean up the conf and then update the `rv-virt` document later.

**Comment by acassis** [comment]
> @acassis, it seems that I've found the way to use virtio-serial as normal console now, so it is no longer output only. I still need time to clean up the conf and then update the `rv-virt` document later.

@yf13 are you using busy await reading (like the LWL console) ? We need to find a way to avoid it, because it eats too much processing time.

**Comment by acassis** [comment]
@yf13 suggestion: instead nsh_aux why don't you use virt_nsh ?

File: drivers/virtio/virtio-serial.c [hunk 2]
```diff
@@ -558,6 +562,15 @@ static int virtio_serial_probe(FAR struct virtio_device *vdev)
       goto err_with_init;
     }
 
+#ifdef CONFIG_VIRTIO_SERIAL_CONSOLE
```
Comment (anchao) [inline drivers/virtio/virtio-serial.c:565]: Why not instantiate the console in the board code?

File: drivers/virtio/virtio-serial.c [hunk 3]
```diff
@@ -595,3 +608,26 @@ int virtio_register_serial_driver(void)
   int ret2 = virtio_register_driver(&g_virtio_rprocserial_driver);
   return ret1 < 0 ? ret1 : ret2;
 }
+
+#ifdef CONFIG_VIRTIO_SERIAL_CONSOLE
```
Comment (anchao) [inline drivers/virtio/virtio-serial.c:612]: ditto

**Comment by yf13** [comment]
@acassis config renamed to `virt_nsh`

But it seems [imx9-sdimage.img](https://github.com/apache/nuttx/actions/runs/10082420944/job/27884276518) has been leaked somewhere? It might block other PR as well. I saw .gitignore already contains a line for that file, but not sure if that is enough?

File: arch/risc-v/src/qemu-rv/qemu_rv_start.c [hunk 4]
```diff
@@ -184,10 +184,14 @@ void qemu_rv_start(int mhartid, const char *dtb)
 
 void riscv_earlyserialinit(void)
 {
+#ifdef HAVE_16550_CONSOLE
```
[the remark this answers is not in the document]
Comment (yf13) [reply arch/risc-v/src/qemu-rv/qemu_rv_start.c:187]: With the `virt_nsh` config, 16550 isn't enabled at all, thus we got error like below if that guard is removed:

```
arch/risc-v/src/qemu-rv/qemu_rv_start.c:79: undefined reference to `u16550_earlyserialinit'
collect2: error: ld returned 1 exit status
```

**Comment by yf13** [comment]
@anchao, thanks for your comments!

I am wondering that serial devices like `uart_16550`, `uart_pl011`, `serial_rtt`, `cddcacm` are quite standard and popular, thus there is no need to treat them as arch specific devices? I feel `virtio-serial` is also quite standard, thus I am treaing it like `uart_16550`.

**Comment by yf13** [comment]
@anchao, thanks for your comments!

I am wondering that serial devices like `uart_16550`, `uart_pl011`, `serial_rtt`, `cddcacm` are quite standard and popular, thus there is no need to treat them as arch specific devices? I feel `virtio-serial` is also quite standard, thus we can treat it like `uart_16550`.

**Comment by yf13** [comment]
@anchao, thanks for your comments!

I am wondering that serial devices like `uart_16550`, `uart_pl011`, `serial_rtt`, `cdcacm` are quite standard, thus there is no need to treat them as arch specific devices? I feel `virtio-serial` is also quite standard, thus we can treat it more like `uart_16550`?

File: arch/risc-v/src/qemu-rv/qemu_rv_start.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (anchao) [reply arch/risc-v/src/qemu-rv/qemu_rv_start.c:187]: You could replace `HAVE_16550_CONSOLE` with `CONFIG_16550_UART`
Refer to: https://github.com/apache/nuttx/blob/master/arch/risc-v/src/k230/k230_start.c#L178-L190

**Comment by anchao** [comment]
> @anchao, thanks for your comments!
> 
> I am wondering that serial devices like `uart_16550`, `uart_pl011`, `serial_rtt`, `cdcacm` are quite standard, thus there is no need to treat them as arch specific devices? I feel `virtio-serial` is also quite standard, thus we can treat it more like `uart_16550`?

@yf13 Your changes are correct, I missed that the tty device is also registered in this driver, please ignore my comments

File: drivers/virtio/Kconfig [hunk 5]
```diff
@@ -74,6 +74,14 @@ if DRIVERS_VIRTIO_SERIAL
 config DRIVERS_VIRTIO_SERIAL_BUFSIZE
 	int "Virtio serial driver buffer size"
 	default 256
+
+config VIRTIO_SERIAL_CONSOLE
```
Comment (anchao) [inline drivers/virtio/Kconfig:78]:
```suggestion
config DRIVERS_VIRTIO_SERIAL_CONSOLE
```

File: drivers/virtio/virtio-serial.c [hunk 6]
```diff
@@ -132,6 +132,10 @@ static const struct uart_ops_s g_virtio_serial_ops =
 
 static int g_virtio_serial_idx = 0;
 
+#ifdef CONFIG_VIRTIO_SERIAL_CONSOLE
+static struct uart_dev_s *g_virtio_console = 0;
```
Comment (anchao) [inline drivers/virtio/virtio-serial.c:136]:
```suggestion
static struct uart_dev_s *g_virtio_console;
```

File: drivers/virtio/virtio-serial.c [hunk 7]
```diff
@@ -558,6 +562,15 @@ static int virtio_serial_probe(FAR struct virtio_device *vdev)
       goto err_with_init;
     }
 
+#ifdef CONFIG_VIRTIO_SERIAL_CONSOLE
+  if (!g_virtio_console)
```
Comment (anchao) [inline drivers/virtio/virtio-serial.c:566]:
```suggestion
  if (g_virtio_console == NULL)
```

File: drivers/virtio/virtio-serial.c [hunk 8]
```diff
@@ -558,6 +562,15 @@ static int virtio_serial_probe(FAR struct virtio_device *vdev)
...
+    {
+      DEBUGVERIFY(uart_register("/dev/console", &priv->udev));
```
Comment (anchao) [inline drivers/virtio/virtio-serial.c:568]: Should the `isconsole` flag be updated in uart driver?
https://github.com/apache/nuttx/blob/master/drivers/serial/uart_16550.c#L1729

**Comment by yf13** [comment]
@anchao, thanks for all the in-line comments! I've updated them,

**Comment by CV-Bowen** [comment]
@yf13 https://github.com/apache/nuttx/pull/12764 Hi, how about support custom the virtio serial device name, and register the console based on the device name (/dev/console).

**Review by acassis — APPROVED** [review]

File: drivers/virtio/virtio-serial.c [hunk 9]
```diff
@@ -558,6 +562,16 @@ static int virtio_serial_probe(FAR struct virtio_device *vdev)
       goto err_with_init;
     }
 
+#ifdef CONFIG_DRIVERS_VIRTIO_SERIAL_CONSOLE
+  if (g_virtio_console == NULL)
+    {
+      DEBUGVERIFY(uart_register("/dev/console", &priv->udev));
```
Comment (xiaoxiang781216) [inline drivers/virtio/virtio-serial.c:568]: it's better to drop this change and utilize this general approach: https://github.com/apache/nuttx/pull/12764

**Review by xiaoxiang781216 — CHANGES_REQUESTED** [review]
let's reuse https://github.com/apache/nuttx/pull/12764

**Comment by anchao** [comment]
These two commits are completely different features, why drop this PR reuse #12764? #12764 was committed as early as "9 Jan 2024", why not submit it earlier ? We need to respect the work of every independent developer, not deny it.

**Review by anchao — APPROVED** [review]

**Comment by acassis** [comment]
> These two commits are completely different features, why drop this PR reuse #12764? #12764 was committed as early as "9 Jan 2024", why not submit it earlier ? We need to respect the work of every independent developer, not deny it.

You are 100% right! "Release early, release often!"

I think we need to respect individual contributor and respect the order the PRs are created.

This one here has priority, please apply this here and ask your developer to adapt his PR later

**Comment by acassis** [comment]
@yf13 please take a look:
```
====================================================================================
Configuration/Tool: rv-virt/ksmp64
2024-07-25 16:00:35
------------------------------------------------------------------------------------
  Cleaning...
HEAD detached at pull/12760/merge
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   boards/risc-v/qemu-rv/rv-virt/configs/virt_nsh/defconfig

no changes added to commit (use "git add" and/or "git commit -a")
  Configuring...
  Building NuttX...
/github/workspace/sources/apps /github/workspace/sources/nuttx
##[debug]Dropping file value '/home/runner/work/nuttx/nuttx/fpu.c'. Path does not exist
Warning: fpu.c:55:8: warning: #warning "FPU test not built; Only available in the flat build (CONFIG_BUILD_FLAT)" [-Wcpp]
   55 | #      warning "FPU test not built; Only available in the flat build (CONFIG_BUILD_FLAT)"
      |        ^~~~~~~
/github/workspace/sources/nuttx
  Normalize rv-virt/ksmp64
```

**Comment by acassis** [comment]
##[debug]Dropping file value '/home/runner/work/nuttx/nuttx/fpu.c'. Path does not exist

**Comment by yf13** [comment]
@acassis, thanks for reminding. It looks fpu.c warning is fine. It was my bad forgetting `tools/refresh.sh` the `virt_nsh` conf after taking @anchao's suggestion.

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**